src/t_hash.c

```diff
@@ -227,31 +227,25 @@ int hashTypeSet(robj *o, sds field, sds value, int flags) {
         if (hashTypeLength(o) > server.hash_max_listpack_entries)
             hashTypeConvert(o, OBJ_ENCODING_HT);
     } else if (o->encoding == OBJ_ENCODING_HT) {
-        dictEntry *de = dictFind(o->ptr,field);
+        dict *ht = o->ptr;
+        dictEntry *de, *existing;
+        sds v;
+        if (flags & HASH_SET_TAKE_VALUE) {
+            v = value;
+            value = NULL;
+        } else {
+            v = sdsdup(value);
+        }
+        de = dictAddRaw(ht, field, &existing);
         if (de) {
-            sdsfree(dictGetVal(de));
-            if (flags & HASH_SET_TAKE_VALUE) {
-                dictGetVal(de) = value;
-                value = NULL;
-            } else {
-                dictGetVal(de) = sdsdup(value);
+            dictSetVal(ht, de, v);
+            if (!(flags & HASH_SET_TAKE_FIELD)) {
+                dictSetKey(ht, de, sdsdup(field));
```

Comment:
don't we have to nullify the `field` in the else of this `if`?

Reply:
Yes, I forgot this. That's my fault. I've updated the code now.